pkg/noderesources/allocatable.go

```diff
@@ -0,0 +1,168 @@
+		if resource.Weight <= 0 {
+			return fmt.Errorf("resource Weight of %v should be a positive value, got %v", resource.Name, resource.Weight)
+		}
+		// No upper bound on weight.
+	}
+	return nil
+}
+
+// Score invoked at the score extension point.
+func (alloc *Allocatable) Score(ctx context.Context, state *framework.CycleState, pod *v1.Pod, nodeName string) (int64, *framework.Status) {
+	nodeInfo, err := alloc.handle.SnapshotSharedLister().NodeInfos().Get(nodeName)
+	if err != nil {
+		return 0, framework.NewStatus(framework.Error, fmt.Sprintf("getting node %q from Snapshot: %v", nodeName, err))
+	}
+
+	// alloc.score favors nodes with least allocatable or most allocatable resources.
+	// It calculates the sum of the node's weighted allocatable resources.
+	//
+	// Note: the returned "score" is negative for least allocatable, and positive for most allocatable.
+	return alloc.score(pod, nodeInfo)
+}
+
+// ScoreExtensions of the Score plugin.
+func (alloc *Allocatable) ScoreExtensions() framework.ScoreExtensions {
+	return alloc
+}
+
+// NewAllocatable initializes a new plugin and returns it.
+func NewAllocatable(allocArgs *runtime.Unknown, h framework.FrameworkHandle) (framework.Plugin, error) {
+	var args NodeResourcesAllocatableArgs
+	if err := framework.DecodeInto(allocArgs, &args); err != nil {
+		return nil, fmt.Errorf("want args to be of type NodeResourcesAllocatableArgs, got %T", allocArgs)
+	}
+
+	if err := validateResources(args.Resources); err != nil {
+		return nil, err
+	}
+
+	resToWeightMap := make(resourceToWeightMap)
+	for _, resource := range args.Resources {
+		resToWeightMap[v1.ResourceName(resource.Name)] = resource.Weight
+	}
+
+	// If resources were not specified, then use the defaults.
+	if len(resToWeightMap) == 0 {
+		resToWeightMap = defaultResourcesToWeightMap
+	}
+
+	// Default mode is least, unless the mode arg is specified.
+	mode := Least
+	if args.Mode != nil {
+		mode = ModeType(*args.Mode)
+		if mode != Least && mode != Most {
+			return nil, fmt.Errorf("invalid mode, got %d", mode)
+		}
+	}
+
+	return &Allocatable{
+		handle: h,
+		resourceAllocationScorer: resourceAllocationScorer{
+			Name:                AllocatableName,
+			scorer:              resourceScorer(resToWeightMap, mode),
+			resourceToWeightMap: resToWeightMap,
+		},
+	}, nil
+}
+
+func resourceScorer(resToWeightMap resourceToWeightMap, mode ModeType) func(resourceToValueMap, resourceToValueMap, bool, int, int) int64 {
+	return func(requested, allocable resourceToValueMap, includeVolumes bool, requestedVolumes int, allocatableVolumes int) int64 {
+		// TODO: consider volumes in scoring.
+		var nodeScore, weightSum int64
+		for resource, weight := range resToWeightMap {
+			resourceScore := score(requested[resource], allocable[resource], mode)
+			nodeScore += resourceScore * weight
+			weightSum += weight
+		}
+		return nodeScore / weightSum
+	}
+}
+
+func score(requested, capacity int64, mode ModeType) int64 {
```

Comment:
seems `requested` is not used at all. If so, you simplify the function signature.

Reply:
Done